### PR TITLE
Remove eval, memoize methodwise, traits, callables, typed caches

### LINKED ISSUE #48: Remove `eval` from at-memoize macro
Right now, the `@memoize` macro creates the cache dictionary using `eval`: 
https://github.com/JuliaCollections/Memoize.jl/blob/1709785afc1eed40d7d35e6c3ebe5ae7e076f57e/src/Memoize.jl#L49-L52
This is generally considered a bad idea in macros, and until recently, also caused an error here because the code was `eval`ing into the Memoize package (see #32).

Right now, using `eval` gives the following behavior:
1. all memoized methods of a given function share the memoization cache
2. if a method is overwritten, the cache is cleared, and any other global resources created/held by the previous version of that method are released: https://github.com/JuliaCollections/Memoize.jl/blob/1709785afc1eed40d7d35e6c3ebe5ae7e076f57e/test/runtests.jl#L257-L266

I made one attempt to simply use a different cache dictionary for each method of a given function (https://github.com/JuliaCollections/Memoize.jl/compare/JuliaCollections:1709785...JuliaCollections:a9170e5).  This mostly works, but as-is causes the test above to fail, because the old cache is not released when the method is overwritten.

There may be an easy way around this, but it's not obvious to me right now.

Thoughts and/or pull requests welcome!

Cc: @cstjean

### LINKED ISSUE #58: LRU example doesn't work
Forgive me if I'm missing something obvious...

```
julia> using Memoize, LRUCache

(@v1.4) pkg> status Memoize LRUCache
Status `~/.julia/environments/v1.4/Project.toml`
  [8ac3fa9e] LRUCache v1.1.0
  [c03570c3] Memoize v0.4.3

julia> @memoize LRU{Tuple{Any,Any},Any}(maxsize=2) function x(a, b)
           println("Running")
           a + b
       end
ERROR: LoadError: MethodError: objects of type LRU{Tuple{Any,Any},Any} are not callable
Stacktrace:
 [1] top-level scope at REPL[12]:1
 [2] eval(::Module, ::Any) at ./boot.jl:331
 [3] @memoize(::LineNumberNode, ::Module, ::Vararg{Any,N} where N) at /Users/fred/.julia/packages/Memoize/V4Yuh/src/Memoize.jl:51
in expression starting at REPL[12]:1
```

### LINKED ISSUE #66: Running @memoize a second time silently does nothing, even if the second call refers to nonexistent names
```julia-repl
julia> @memoize LRU(maxsize=10) f(x) = x + 1
f (generic function with 1 method)

julia> @memoize NonexistentCacheType f(x) = x + 1
f (generic function with 1 method)

julia> f(3)
4
```

I would expect an error (or a loud warning message) on that second call.  At the very least, I'd want a loud warning in the documentation.

### LINKED ISSUE #68: Saving cache to file and reading from it
Hi, I am a new user of memoize and need some help.

As I can see, memoize can use dictionary as a cache which I can probably write to a file later. Can I read this dictionary later and load it into the cache?

I checked the asked questions but it seems this feature doesn't exist(?)

### PR DESCRIPTION
**Pull request by peterahrens**
This PR picks up where #59 left off (thanks @cstjean for all your work on this repo!). If we remove `eval` and use one cache per method, then we need to be able to look up caches by their methods. Types can't be hashed by equivalence classes (otherwise I think Julia dispatch might be faster, consider `hash(Val{A} where {A}) != hash(Val{B} where B)`). Therefore, we need to use Julia's method lookup to find methods to find their cache. We use a global dictionary in the Memoize.jl module to map `Method` objects to their caches. Since this was already a big change, this PR also makes several related changes.

The key changes made by this PR
- Removes `eval`. (fixes #48)
- Caches are associated with methods, rather than functions. (fixes #66)
- Supports callable types, callable objects, and lambdas (Doesn't support inner constructors, I think)
- Supports trait memoization (i.e. arguments of the form `(::T)` can now be memoized, and they use `T` as their key)
- Removed syntactic special casing of dictionary construction, in favor of the simply supplying an expression which constructs the cache. (fixes #58)
- Defines local variables `__Key__` and `__Value__` in the cache constructor expression scope, to support more specific dictionary types.
- Only empties caches when methods are overwritten. (fixes #68)

In summary, now you can do:
```
struct F{A}
	a::A
end
@memoize Dict{__Key__, __Value__}() function (::F{A})(b, ::C) where {A, C}
	println("Running")
	(A, b, C)
end
```

### CODE REVIEW

File: src/Memoize.jl [hunk 1]
```diff
@@ -1,89 +1,196 @@
...
+        if meth isa Method
+            return meth::Method
+        else
+            meth = meth.func
+            return meth::Method
+        end
+    end
+end
+
+const _brain = Dict()
```
Comment (cstjean) [inline src/Memoize.jl:18]: A global like this is problematic because of precompilation. Eg. if you have a package with

```julia
module X
@memoize foo(x) = x
end
```

I believe you'll find that after `using Memoize, X`, `Memoize.brain` is empty.

**Comment by cstjean** [comment]
Hi Peter, this is an impressive list of fixes! I love all the tests and README additions.

For cache invalidation, from an API perspective, I believe we only really need:

1. A way to clear all caches for a function
2. Automatically clear the specific method's cache when it's redefined

I like the direction you took with `_memories(...)`. To make it precompilation-friendly, we might need to get rid of the global dict. I'm not sure how to do it, but consider expanding `@memoize foo(x::Int, y) = ...` into

```julia
...
global var"__memoize_cache_foo(x::Int,y)" = Dict()
...
```

then on redefinition, we can just `empty!(var"__memoize_cache_foo(x::Int,y)")`, achieving goal 2. For goal 1, we can take the `method` list as you did, find each method's definition module, and look at `names(that_module)` for all variables starting with `__memoize_cache_foo`. Then we empty all of these caches. This is not elegant, but I believe it would work.

Cache emptying is the first problem to solve. If you're interested in implementing that (or another precompilation-compatible solution), I would appreciate if that was a small, self-contained PR starting from #59 (you can choose the target on github). 

It's really best to split big PRs into chunks. It's so much easier to review and merge a small PR!

**Comment by peterahrens** [comment]
Good catch regarding precompilation. I think my most recent commits fix this using something similar to your suggestion, but using only one variable holding the dictionary instead of one variable per method.

I'll consider splitting this PR over the next few weeks/months.

**Comment by peterahrens** [comment]
Reflecting a little bit on the proposed solution, and other potential ways to simplify #71:

> I like the direction you took with _memories(...). To make it precompilation-friendly, we might need to get rid of the global dict. I'm not sure how to do it, but consider expanding `@memoize foo(x::Int, y) = ...` into `global var"__memoize_cache_foo(x::Int,y)" = Dict()`

>then on redefinition, we can just empty!(var"__memoize_cache_foo(x::Int,y)"), achieving goal 2. For goal 1, we can take the method list as you did, find each method's definition module, and look at names(that_module) for all variables starting with __memoize_cache_foo. Then we empty all of these caches. This is not elegant, but I believe it would work.

I'm not convinced this approach will work. It doesn't detect overwrites when variables are named differently. For example, the string `"__memoize_cache_foo(x::Int,y)"` will not equal the string `"__memoize_cache_foo(x::Int,z)"` even when the second method would overwrite the first. It also spuriously detects overwrites when the types themselves are variables that need resolution, such as `T = Int; @memoize foo(::T) = 1; T = Bool; @memoize foo(::T) = 2`.

Defining a method overwrites a previous method precisely when their type signatures match, so it seems best to use the type signatures (e.g. `Tuple{typeof(foo), Int, Any}`) of the methods as keys to associate with a cache. Unfortunately, while it is easy to check if two types are equivalent (one can use `==`), it seems that there is no straightforward way to _find_ equivalent types. I'm imagining a few options.

1. We use a list to look up caches. When defining a method, we walk the whole list to find signatures of potential methods that might be overwritten, invalidate and remove that cache if found, then add the signature and cache of our new method to the list.
2. We use a sentinel method to look up caches. When defining `foo(x::Int,y)`, we also define `find_the_cache(::Type{Tuple{typeof(foo), Int, Any}}) = Dict()`. Of course, we also define `find_the_cache(::Any) = nothing`, so that we can call `find_the_cache` to determine if our method has been defined before. This could greatly simplify the code. Unfortunately, if `find_the_cache` is to be a global method, this would put us back in the world of needing to call eval.
3. #70 and #71 use Julia introspection to find the method that would be overwritten, if any. Since the signature of that method matches the known signature exactly, we can hash types based on their signature with a dictionary that we store in each module.

I think that the first solution would likely be easier to read than #71, and I'm not sure anyone will notice the slowdown of linear search versus whatever julia uses internally to solve the "find equivalent type" problem. We might also simplify #71 a bit by calling a more standard form of `which.`

**Comment by peterahrens** [comment]
> To make it precompilation-friendly, we might need to get rid of the global dict.
If this is a concern because there was only one global dict in the Memoize package, #71 has been amended to use a dictionary in each package that uses memoize.

If, however, this was a concern because of the interactions between precompilation and rehashing dictionaries; whatever problems we have with the global dictionary that Memoize creates, we'll also have with the dictionaries the global dict is attempting to store, and that might be a separate issue.

**Comment by cstjean** [comment]
> I'm not convinced this approach will work. It doesn't detect overwrites when variables are named differently. For example, the string "__memoize_cache_foo(x::Int,y)" will not equal the string "__memoize_cache_foo(x::Int,z)" even when the second method would overwrite the first.

That's true, but to me that's basically a negligible problem, affecting a vanishingly small number of real-world scenarios.

> It also spuriously detects overwrites when the types themselves are variables that need resolution, such as T = Int; @memoize foo(::T) = 1; T = Bool; @memoize foo(::T) = 2.

I can't think of any code I've ever seen that would do something like that. Can you? People might write

```julia
for T in [:Int, :Vector]
     @eval @memoize f(::$T) = ...
end
```

but that would work fine, because the $ happens before the `@memoize` expansion.

**Comment by cstjean** [comment]
> We use a list to look up caches. When defining a method, we walk the whole list to find signatures of potential methods that might be overwritten, invalidate and remove that cache if found, then add the signature and cache of our new method to the list.

Since the list is global, wouldn't you be stuck with the same precompilation problems as before?

> We use a sentinel method to look up caches. When defining foo(x::Int,y), we also define find_the_cache(::Type{Tuple{typeof(foo), Int, Any}}) = Dict(). Of course, we also define find_the_cache(::Any) = nothing, so that we can call find_the_cache to determine if our method has been defined before. This could greatly simplify the code. 

I like that... if it works! 🙂 Memoize.jl is such a julia puzzle to figure out.

> Unfortunately, if find_the_cache is to be a global method, this would put us back in the world of needing to call eval.

I don't understand. For local variables, I don't think we need to support explicit user-side cache invalidation, so this looks fine to me (as long as the cache can be GC'ed once the function returns)

> If this is a concern because there was only one global dict in the Memoize package, #71 has been amended to use a separate dictionary in each package that uses memoize.

👍 

> If, however, this was a concern because of the interactions between precompilation and rehashing dictionaries; 

That's new to me, what are those interactions?

**Comment by peterahrens** [comment]
After attempting to implement 2, I suspect it might be impossible to avoid introspection: We need to find the module in which the overwritten method was defined. I can't think of a way to do it without calling `which`, meaning that 1. or 2. aren't any simpler than 3. I think your variable-based approach might also require introspection to determine which module stores the variable. As long as we need to look up the method, we might as well use its type signature to correctly detect overwrites and avoid name-dependent behavior.

**Comment by cstjean** [comment]
Yes, with what I suggested we need introspection too, as I said:

> For goal 1, we can take the method list as you did, find each method's definition module, and look at names(that_module) for all variables starting with __memoize_cache_foo. Then we empty all of these caches. This is not elegant, but I believe it would work

> As long as we need to look up the method, we might as well use its type signature to correctly detect overwrites and avoid name-dependent behavior.

Like I said, I don't think the name-dependent behaviour is problematic, but if you can make it work with similar complexity, I will be happy to do the right thing and use types.
